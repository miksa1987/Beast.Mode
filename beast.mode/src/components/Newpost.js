import React, { useState } from 'react'
import { connect } from 'react-redux'
import { Form, TextArea, Button, Checkbox } from 'semantic-ui-react'
import parser from '../service/parser'
import communicationService from '../service/communication'
import { addToFeed } from '../reducers/feedReducer'

const elementStyle = {
  align: 'center',
  margin: 'auto',
  minWidth: '95%',
  maxWidth: '95%',
  paddingTop: '5px',
  paddingLeft: '5px',
  flexGrow: '1',
  flexShrink: '1',
  flexBasis: '95%'
}

const buttonStyle = {
  margin: '5px 5px 5px 5px'
}

const Newpost = (props) => {
  const [isWorkout, setIsWorkout] = useState(false)
  const [textContent, setTextContent] = useState('')
  const [file, setFile] = useState('')

  const style = {
    resize: 'none'
  }

  const changeText = (event) => {
    setTextContent(event.target.value)
    console.log(textContent)
    if (parser.isWorkout(textContent)) {
      setIsWorkout(true)
    } else {
      setIsWorkout(false)
    }
    console.log(isWorkout)
  }

  const workoutToggleChange = (event, data) => setIsWorkout(data.checked)

  const post = async (event) => {
    event.preventDefault()
    let post = {
      content: textContent,
      picture: '',
      user: props.currentUser.id,
      likes: 0,
      comments: []
    }

    const data = new FormData()
    data.append('content', textContent)
    data.append('image', file)
    data.append('user', props.currentUser.id)
    data.append('likes', 0)
    isWorkout ? data.append('type', 'workout') : data.append('type', 'post')
    isWorkout ? post.type = 'workout' : post.type = 'post'
    console.log(data)

    const header = {
      'content-type': 'multipart/form-data'
    }
    let newPost = {}

    if(post.type === 'post') newPost =  await communicationService.post('/posts/new', data, header)
    if(post.type === 'workout') newPost = await communicationService.post('/workouts/new', data, header)
    if(post.type === 'doneworkout') newPost = await communicationService.post('/doneworkouts/new', data, header)

    props.addToFeed(newPost)
  }

  return ( <div style={elementStyle}>
      <Form onSubmit={post}>
        <input type='file' onChange={({ target }) => setFile(target.files[0])} />
        <TextArea style={style} name='post' onChange={changeText} rows={8} placeholder='What have you done?! (tip: you can use hashtags!)' />
        { isWorkout ? <><Checkbox toggle name='workoutToggle' onChange={workoutToggleChange} /><strong>Did it!</strong></> : null }
        <Button style={buttonStyle}>Got picture?</Button>
        <Button type='submit' style={buttonStyle}>Post!</Button>
      </Form>
  </div> )
}

const mapStateToProps = (state) => {
  return { currentUser: state.currentUser }
}

export default connect(mapStateToProps, { addToFeed })(Newpost)