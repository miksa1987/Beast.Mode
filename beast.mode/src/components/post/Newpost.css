@media (max-width: 600px) {
  .button-style {
    margin-top: 10px;
    padding: 0px;
  }
}

@media (min-width: 620px) {
  .button-style {
    margin-top: 10px;
  }
}
.newpost-component {
  padding: 10px 10px 10px 10px;
  border: 1px solid #dddddd;
  border-radius: 4px;
  background-color: #ffffff;
}

.input-style {
  margin-bottom: 10px;
}

.button-style {
  margin-top: 10px;
}

.mobile {
  animation: fade-in 1.5s;
  padding: 10px 10px 10px 10px;
  border: 1px solid #dddddd;
  border-radius: 4px;
  background-color: #ffffff;
}

.fileContainer {
  overflow: hidden;
  position: relative;
}

.fileContainer [type=file] {
  cursor: inherit;
  display: block;
  font-size: 999px;
  filter: alpha(opacity=0);
  min-height: 100%;
  min-width: 100%;
  opacity: 0;
  position: absolute;
  right: 0;
  text-align: right;
  top: 0;
}