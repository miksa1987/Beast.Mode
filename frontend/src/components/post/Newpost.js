import React, { useState } from 'react'
import { connect } from 'react-redux'
import { Form, TextArea, Button, Image } from 'semantic-ui-react'
import communicationService from '../../service/communication'
import { addNewToFeed } from '../../reducers/feedReducer'
import { addWorkout } from '../../reducers/workoutsReducer'
import useField from '../../hooks/useField'
import FileInput from '../universal/FileInput'
import FlexDivColumn from '../universal/FlexDivColumn'
import FlexDivRow from '../universal/FlexDivRow'
import Animation from '../universal/Animation'

const Newpost = (props) => {
  const [isWorkout, setIsWorkout] = useState(props.isWorkout)
  const [didWorkout, setDidWorkout] = useState(false)
  const [text, resetText] = useField('text')
  const [file, setFile] = useState('')
  const [image, setImage] = useState('')

  const workoutTip = `Write your workout like:
    WORKOUT TITLE OR DESCRIPTION
    5x5 pull ups
    5x10 push ups\n
    or\n
    WORKOUT TITLE OR DESCRIPTION
    5 pull ups
    10 push ups
    5 rounds`

  const post = async (event) => {
    event.preventDefault()

    let post = {
      content: text.value,
      picture: image,
      user: props.currentUser.id,
      likes: 0,
      type: 'post'
    }

    if (isWorkout) {
      if (didWorkout) post.type = 'doneworkout'
      else post.type = 'workout'
    }

    let newPost = {}
    if(!isWorkout) props.addNewToFeed(post)
    if(isWorkout && !didWorkout) newPost = await communicationService.post('/workouts/new', post)
    if(isWorkout && didWorkout) props.addNewToFeed(post)

    isWorkout && props.addWorkout(newPost)

    resetText()
    props.setShowNewpost && props.setShowNewpost(false)
  }
  
  return (
    <Animation>
      <FlexDivColumn>

        <textarea id='post-textarea' style={{ resize: 'none' }} rows={6} {...text} 
        placeholder={isWorkout && workoutTip} />

        <FlexDivRow>
          <Image src={image ? image 
            : 'https://react.semantic-ui.com/images/wireframe/image.png'} size='mini' />
          <FileInput file={file} setFile={setFile} setImage={setImage} />
        </FlexDivRow>

        <FlexDivRow>
          <Button compact data-testid='postbutton' id='postbutton' color='red' className='button-style' type='submit'>Post</Button>

          {!props.isWorkout && <Button.Group>
            <Button compact id='updatebutton' data-testid='updatebutton' className='button-style' type='button' 
              color={!isWorkout ? 'red' : 'black'} onClick={() => setIsWorkout(false)}>Update</Button>

            <Button compact id='workoutbutton' data-testid='workoutbutton' className='button-style' type='button' 
              color={isWorkout ? 'red' : 'black'} onClick={() => setIsWorkout(true)}>Workout</Button>
          </Button.Group>}

              {` `}

          {props.setShowNewpost && <Button compact data-testid='cancelbutton' id='cancelbutton' className='button-style' 
            color='red' floated='right' onClick={() => props.setShowNewpost(false)}>Cancel</Button>}

          {isWorkout && <Button compact id='didworkoutbutton' data-testid='didworkoutbutton' className='button-style' type='button' 
            color={didWorkout ? 'red' : 'black'} floated='right' onClick={() => setDidWorkout(!didWorkout)}>Did it?</Button>}
        </FlexDivRow>

      </FlexDivColumn>
    </Animation>
  )
}

const mapStateToProps = (state) => {
  return { 
    currentUser: state.currentUser 
  }
}

export default connect(mapStateToProps, { addNewToFeed, addWorkout })(Newpost)