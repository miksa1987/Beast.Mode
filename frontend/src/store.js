import { createStore, combineReducers, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'

import currentUserReducer from './reducers/currentUser'
import feedReducer from './reducers/feedReducer'
import workoutsReducer from './reducers/workoutsReducer'
import usersReducer from './reducers/usersReducer'
import doneWorkoutsReducer from './reducers/doneWorkoutsReducer'
import currentUserPostsReducer from './reducers/currentUserPosts'
import currentWorkoutReducer from './reducers/currentWorkout'
import currentProfileReducer from './reducers/currentProfile'
import currentPostReducer from './reducers/currentPost'
import notificationReducer from './reducers/notificationReducer'
import searchResultsReducer from './reducers/searchResultsReducer'

const reducer = combineReducers({
  currentUser: currentUserReducer,
  currentUserPosts: currentUserPostsReducer,
  currentWorkout: currentWorkoutReducer,
  currentProfile: currentProfileReducer,
  currentPost: currentPostReducer,
  users: usersReducer,
  feed: feedReducer,
  workouts: workoutsReducer,
  doneWorkouts: doneWorkoutsReducer,
  notification: notificationReducer,
  search: searchResultsReducer
})

const store = createStore(reducer, applyMiddleware(thunk))

export default store