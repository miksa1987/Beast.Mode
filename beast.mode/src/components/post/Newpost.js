import React, { useState } from 'react'
import { connect } from 'react-redux'
import { Form, TextArea, Button, Image } from 'semantic-ui-react'
import communicationService from '../../service/communication'
import { addNewToFeed } from '../../reducers/feedReducer'
import { addWorkout } from '../../reducers/workoutsReducer'
import useField from '../../hooks/useField'
import FileInput from '../universal/FileInput'
import './Newpost.css'
import '../Animation.css'

const Newpost = (props) => {
  const [isWorkout, setIsWorkout] = useState(props.isWorkout)
  const [didWorkout, setDidWorkout] = useState(false)
  const [text, resetText] = useField('text')
  const [file, setFile] = useState('')
  const [image, setImage] = useState('')

  const post = async (event) => {
    event.preventDefault()

    let post = {
      content: text.value,
      picture: image,
      user: props.currentUser.id,
      likes: 0,
      type: 'post'
    }

    if (isWorkout) {
      if (didWorkout) post.type = 'doneworkout'
      else post.type = 'workout'
    }

    let newPost = {}
    if(!isWorkout) props.addNewToFeed(post)
    if(isWorkout && !didWorkout) newPost = await communicationService.post('/workouts/new', post)
    if(isWorkout && didWorkout) props.addNewToFeed(post)

    isWorkout ? props.addWorkout(newPost) : console.log('HII')

    resetText()
    props.setShowNewpost && props.setShowNewpost(false)
  }
  
  return ( <div className='newpost-component fade-in-fast'>
    <strong>Create new</strong>
    <Form onSubmit={post}>
      <TextArea id='post-textarea' style={{ resize: 'none' }} rows={6} {...text} />
      
      <table className='full-width'>
        <tbody>
          <tr>
            <td>
              <Image src={image ? image 
              : 'https://react.semantic-ui.com/images/wireframe/image.png'} size='mini' />
            </td>
            <td>
              <FileInput file={file} setFile={setFile} setImage={setImage} />
            </td>
          </tr>
        </tbody>
      </table>
  
      <Button compact id='postbutton' primary className='button-style' type='submit'>Post</Button>

      {!props.isWorkout && <Button.Group>
        <Button compact id='updatebutton' className='button-style' type='button' color={!isWorkout ? 'blue' : 'black'} 
          onClick={() => setIsWorkout(false)}>Update</Button>
        <Button compact id='workoutbutton' className='button-style' type='button' color={isWorkout ? 'blue' : 'black'} 
          onClick={() => setIsWorkout(true)}>Workout</Button>
      </Button.Group>}

      {` `}

      {props.setShowNewpost && <Button compact id='didworkoutbutton' className='button-style' color='blue' floated='right' 
      onClick={() => props.setShowNewpost(false)}>Cancel</Button>}
      {isWorkout && <Button compact id='didworkoutbutton' className='button-style' type='button' color={didWorkout ? 'blue' : 'black'}
        floated='right' onClick={() => setDidWorkout(!didWorkout)}>Did it?</Button>}
    </Form>
  </div> )
}

const mapStateToProps = (state) => {
  return { 
    currentUser: state.currentUser 
  }
}

export default connect(mapStateToProps, { addNewToFeed, addWorkout })(Newpost)