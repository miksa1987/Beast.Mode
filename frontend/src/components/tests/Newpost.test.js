import React from 'react'
import { Provider } from 'react-redux'
import '@testing-library/jest-dom/extend-expect'
import { render, cleanup, fireEvent } from '@testing-library/react'
import Newpost from '../post/Newpost'
import store from '../../store'

afterEach(cleanup)

test('Header is rendered', () => {
  const component = render(<Provider store={store}><Newpost /></Provider>)
  expect(component.container).toHaveTextContent('Create new')
})

test('Can set to workout type and did it button appears', () => {
  const component = render(<Provider store={store}><Newpost /></Provider>)
  const workoutButton = component.getByTestId('workoutbutton')
  fireEvent.click(workoutButton)
  const didIt = component.getByTestId('didworkoutbutton')
  
  expect(didIt).toBeDefined()
})

test('When isWorkout is set, post type buttons are not shown', () => {
  const setShowNewPost = jest.fn()
  const component = render(<Provider store={store}><Newpost isWorkout={true} setShowNewpost={setShowNewPost} /></Provider>)
  const postButton = component.getByTestId('postbutton')
  const didIt = component.getByTestId('didworkoutbutton')
  const cancel = component.getByTestId('cancelbutton')
  
  expect(postButton).toBeDefined()
  expect(didIt).toBeDefined()
  expect(cancel).toBeDefined()
})

test('When isWorkout is set and cancel clicked setShowNewpost is called once', () => {
  const setShowNewPost = jest.fn()
  const component = render(<Provider store={store}><Newpost isWorkout={true} setShowNewpost={setShowNewPost} /></Provider>)
  const cancel = component.getByTestId('cancelbutton')
  fireEvent.click(cancel)
  
  expect(setShowNewPost.mock.calls.length).toBe(1)
})