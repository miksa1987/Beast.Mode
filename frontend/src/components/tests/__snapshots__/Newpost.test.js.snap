// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is rendered correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-htpNat dAVcta"
      >
        <form
          data-testid="newpost-component"
        >
          <div
            class="sc-bdVaJa dJrvce"
          >
            <textarea
              id="post-textarea"
              rows="6"
              style="resize: none;"
              type="text"
            />
            <div
              class="sc-bwzfXH eGZimR"
            >
              <img
                class="ui mini image"
                src="https://react.semantic-ui.com/images/wireframe/image.png"
              />
              <div>
                <input
                  id="real-input"
                  type="file"
                />
                <button
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="image icon"
                  />
                  Upload a picture
                </button>
              </div>
            </div>
            <div
              class="sc-bwzfXH eGZimR"
            >
              <button
                class="button-style"
                data-testid="postbutton"
                id="postbutton"
                style="background-color: orange;"
                type="submit"
              >
                Post
              </button>
              <div>
                <button
                  class="button-style"
                  data-testid="updatebutton"
                  id="updatebutton"
                  style="background-color: orange;"
                  type="button"
                >
                  Update
                </button>
                <button
                  class="button-style"
                  data-testid="workoutbutton"
                  id="workoutbutton"
                  style="background-color: lightgrey;"
                  type="button"
                >
                  Workout
                </button>
              </div>
               
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-htpNat dAVcta"
    >
      <form
        data-testid="newpost-component"
      >
        <div
          class="sc-bdVaJa dJrvce"
        >
          <textarea
            id="post-textarea"
            rows="6"
            style="resize: none;"
            type="text"
          />
          <div
            class="sc-bwzfXH eGZimR"
          >
            <img
              class="ui mini image"
              src="https://react.semantic-ui.com/images/wireframe/image.png"
            />
            <div>
              <input
                id="real-input"
                type="file"
              />
              <button
                type="button"
              >
                <i
                  aria-hidden="true"
                  class="image icon"
                />
                Upload a picture
              </button>
            </div>
          </div>
          <div
            class="sc-bwzfXH eGZimR"
          >
            <button
              class="button-style"
              data-testid="postbutton"
              id="postbutton"
              style="background-color: orange;"
              type="submit"
            >
              Post
            </button>
            <div>
              <button
                class="button-style"
                data-testid="updatebutton"
                id="updatebutton"
                style="background-color: orange;"
                type="button"
              >
                Update
              </button>
              <button
                class="button-style"
                data-testid="workoutbutton"
                id="workoutbutton"
                style="background-color: lightgrey;"
                type="button"
              >
                Workout
              </button>
            </div>
             
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
